presto-expressions/src/test/java/com/facebook/presto/expressions/translator/TestRowExpressionTranslator.java

```diff
@@ -0,0 +1,182 @@
+
+import com.facebook.presto.metadata.FunctionManager;
+import com.facebook.presto.metadata.Metadata;
+import com.facebook.presto.metadata.MetadataManager;
+import com.facebook.presto.spi.ColumnHandle;
+import com.facebook.presto.spi.function.FunctionMetadata;
+import com.facebook.presto.spi.function.ScalarFunction;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.function.StandardFunctionResolution;
+import com.facebook.presto.spi.relation.CallExpression;
+import com.facebook.presto.spi.relation.ConstantExpression;
+import com.facebook.presto.spi.relation.DeterminismEvaluator;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.SpecialFormExpression;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.spi.type.StandardTypes;
+import com.facebook.presto.sql.TestingRowExpressionTranslator;
+import com.facebook.presto.sql.planner.TypeProvider;
+import com.facebook.presto.sql.relational.FunctionResolution;
+import com.facebook.presto.sql.relational.RowExpressionDeterminismEvaluator;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import org.testng.annotations.Test;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.expressions.translator.FunctionTranslator.buildFunctionTranslator;
+import static com.facebook.presto.expressions.translator.RowExpressionTreeTranslator.translateWith;
+import static com.facebook.presto.expressions.translator.TranslatedExpression.untranslated;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.sql.planner.iterative.rule.test.PlanBuilder.expression;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptyMap;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+
+public class TestRowExpressionTranslator
+{
+    private static final Metadata METADATA = MetadataManager.createTestMetadataManager();
+
+    private final FunctionManager functionManager;
+    private final StandardFunctionResolution functionResolution;
+    private final DeterminismEvaluator determinismEvaluator;
+    private final TestingRowExpressionTranslator sqlToRowExpressionTranslator;
+
+    public TestRowExpressionTranslator()
+    {
+        this.functionManager = METADATA.getFunctionManager();
+        this.functionResolution = new FunctionResolution(functionManager);
+        this.determinismEvaluator = new RowExpressionDeterminismEvaluator(functionManager);
+        this.sqlToRowExpressionTranslator = new TestingRowExpressionTranslator(METADATA);
+    }
+
+    @Test
+    public void testEndToEndFunctionTranslation()
+    {
+        String untranslated = "LN(bitwise_and(1, col1))";
```

Comment:
We also might want to cover:
some types of constant that is not translatable.

Reply:
what type of constant would that be? untranslatable constants wouldnt make it past expression creation no? since the type provider entry will be missing?